components/external_storage/src/s3.rs

```diff
@@ -0,0 +1,209 @@
+use rusoto_credential::StaticProvider;
+use rusoto_s3::*;
+
+use super::ExternalStorage;
+use kvproto::backup::S3 as Config;
+
+/// S3 compatible storage
+#[derive(Clone)]
+pub struct S3Storage {
+    config: Config,
+    client: S3Client,
+}
+
+impl S3Storage {
+    /// Create a new S3 storage for the given config.
+    pub fn new(config: &Config) -> Result<S3Storage> {
+        // This can greatly improve performance dealing with payloads greater
+        // than 100MB. See https://github.com/rusoto/rusoto/pull/1227
+        // for more information.
+        let mut http_config = HttpConfig::new();
+        http_config.read_buf_size(1024 * 1024 * 2);
+        let http_dispatcher = HttpClient::new_with_config(http_config).unwrap();
+
+        S3Storage::with_request_dispatcher(config, http_dispatcher)
+    }
+
+    fn with_request_dispatcher<D>(config: &Config, dispatcher: D) -> Result<S3Storage>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+        D::Future: Send,
+    {
+        if config.bucket.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidInput, "missing bucket name"));
+        }
+        if config.prefix.contains("/") {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                format!("multi-prefix is not allowed: {}", config.prefix),
+            ));
+        }
+        if config.access_key.is_empty() {
+            return Err(Error::new(ErrorKind::InvalidInput, "missing access_key"));
+        }
+        if config.secret_access_key.is_empty() {
+            return Err(Error::new(
+                ErrorKind::InvalidInput,
+                "missing secret_access_key",
+            ));
+        }
+        let static_cred = StaticProvider::new(
+            config.access_key.clone(),
+            config.secret_access_key.clone(),
+            None, /* token */
+            None, /* valid_for */
+        );
+        let region = if config.endpoint.is_empty() {
+            config.region.parse::<region::Region>().map_err(|e| {
+                Error::new(
+                    ErrorKind::InvalidInput,
+                    format!("invalid region format {}: {}", config.region, e),
+                )
+            })?
+        } else {
+            region::Region::Custom {
+                name: config.region.clone(),
+                endpoint: config.endpoint.clone(),
+            }
+        };
+        Ok(S3Storage {
+            config: config.clone(),
+            client: S3Client::new_with(dispatcher, static_cred, region),
+        })
+    }
+
+    fn maybe_prefix_key(&self, key: &str) -> String {
+        if !self.config.prefix.is_empty() {
+            return format!("{}/{}", self.config.prefix, key);
+        }
+        key.to_owned()
+    }
+}
+
+impl ExternalStorage for S3Storage {
+    fn write(&self, name: &str, reader: &mut dyn Read) -> Result<()> {
+        let mut content = vec![];
+        reader.read_to_end(&mut content)?;
```

Comment:
Please don't read the entire file into memory.

Reply:
Mind if I do it in another PR?

Reply:
It's hard to say yes when this is explicitly called out in https://github.com/pingcap/br/issues/89#issuecomment-563705345 😅

Reply:
Not mean to be an execute, I thought the change is more complex like using multipart upload API. Turns out we only want to avoid memcopy from `reader` to `body` of the request?

Reply:
> Turns out we only want to avoid memcopy from reader to body of the request?

Yes, because the size of SST file can be large. Basically you need to find a way to [construct an `impl Stream<Item = Bytes> + Send + 'static`](https://docs.rs/rusoto_core/0.42.0/rusoto_core/struct.ByteStream.html) out of the `reader: &mut dyn Read` (maybe `reader` need to be changed to `AsyncRead`; or maybe you just spawn a task right here).